egs/aishell/s10/chain/tdnnf_layer.py

```diff
@@ -0,0 +1,319 @@
+        self.batchnorm1 = nn.BatchNorm1d(num_features=big_dim)
+        self.linear = OrthonormalLinear(dim=big_dim,
+                                        bottleneck_dim=small_dim,
+                                        time_stride=0)
+        self.batchnorm2 = nn.BatchNorm1d(num_features=small_dim)
+
+    def forward(self, x):
+        # x is [N, C, T]
+        x = x.permute(0, 2, 1)
+
+        # at this point, x is [N, T, C]
+
+        x = self.affine(x)
+        x = F.relu(x)
+
+        # at this point, x is [N, T, C]
+
+        x = x.permute(0, 2, 1)
+
+        # at this point, x is [N, C, T]
+
+        x = self.batchnorm1(x)
+
+        x = self.linear(x)
+
+        x = self.batchnorm2(x)
+
+        return x
+
+    def constrain_orthonormal(self):
+        self.linear.constrain_orthonormal()
+
+
+class FactorizedTDNN(nn.Module):
+    '''
+    This class implements the following topology in kaldi:
+      tdnnf-layer name=tdnnf2 $tdnnf_opts dim=1024 bottleneck-dim=128 time-stride=1
+
+    References:
+        - http://danielpovey.com/files/2018_interspeech_tdnnf.pdf
+        - ConstrainOrthonormalInternal() from
+          https://github.com/kaldi-asr/kaldi/blob/master/src/nnet3/nnet-utils.cc#L982
+    '''
+
+    def __init__(self,
+                 dim,
+                 bottleneck_dim,
+                 time_stride,
+                 conv_stride,
+                 bypass_scale=0.66):
+        super().__init__()
+
+        assert conv_stride in [1, 3]
+        assert abs(bypass_scale) <= 1
+
+        self.bypass_scale = bypass_scale
+
+        self.conv_stride = conv_stride
+
+        # linear requires [N, C, T]
+        self.linear = OrthonormalLinear(dim=dim,
+                                        bottleneck_dim=bottleneck_dim,
+                                        time_stride=time_stride)
+
+        # affine requires [N, C, T]
+        # WARNING(fangjun): we do not use nn.Linear here
+        # since we want to use `stride`
+        self.affine = nn.Conv1d(in_channels=bottleneck_dim,
+                                out_channels=dim,
+                                kernel_size=1,
+                                stride=conv_stride)
+
+        # batchnorm requires [N, C, T]
+        self.batchnorm = nn.BatchNorm1d(num_features=dim)
+
+    def forward(self, x):
+        # input x is of shape: [batch_size, feat_dim, seq_len] = [N, C, T]
+        assert x.ndim == 3
+
+        # save it for skip connection
+        input_x = x
+
+        x = self.linear(x)
+
+        # at this point, x is [N, C, T]
+
+        x = self.affine(x)
+
+        # at this point, x is [N, C, T]
+
+        x = F.relu(x)
+
+        # at this point, x is [N, C, T]
+
+        x = self.batchnorm(x)
+
+        # at this point, x is [N, C, T]
+
+        # TODO(fangjun): implement GeneralDropoutComponent in PyTorch
+
+        if self.linear.kernel_size == 3:
+            x = self.bypass_scale * input_x[:, :, 1:-1:self.conv_stride] + x
```

Comment:
shouldn't this be `c:-c:c` rather than `1:-1:c`, where `c` is self.conv_stride?

Reply:
Suppose the time_stride is 1 and the conv_stride is 1.

If the input time index is
```
0 1 2 3 4 5 6
```

After `self.linear`, the time index will be

```
1 2 3 4 5
```
since the kernel shape is `[-1, 0, 1]` (time_stride == 1)

After `self.affine`, the time index is still 

```
1 2 3 4 5
```

The index of `input[1:-1:self.conv_stride]` is `[1, 2, 3, 4, 5]` which matches 
the output of `self.affine`.

----
It is assumed that

- `time_stride == 1`, `conv_stride == 1`

or

- `time_stride == 0`, `conv_stride == 3`

So `c:-c:c` is equivalent to `1:-1:c` when `time_stride==1` and `conv_stride == 1`.

Reply:
I don't think it should be called time_stride here.  Perhaps in the original Kaldi code it wasn't super clear but when implemented as convolution it gets very confusing.   Better to make (stride, kernel_size) the parameters and have them be (1, 3), (1, 3), ... (3, 3), (1, 1), (1, 3), (1, 3) ...
In any case, please revert other aspects of the implementation to more similar to the way it was before and start doing experiments with that.  I don't see much point starting from such a strange starting point.  (i.e. the way the code is right now).

Reply:
I agree.

I also find them confusing but I wrote it this way to follow the naming style in Kaldi.

I'll change them now.